types/underscore/index.d.ts

```diff
@@ -729,17 +711,21 @@ declare module _ {
         size<T>(list: _.Collection<T>): number;
 
         /**
-        * Split array into two arrays:
-        * one whose elements all satisfy predicate and one whose elements all do not satisfy predicate.
-        * @param array Array to split in two.
-        * @param iterator Filter iterator function for each element in `array`.
-        * @param context `this` object in `iterator`, optional.
-        * @return Array where Array[0] are the elements in `array` that satisfies the predicate, and Array[1] the elements that did not.
-        **/
-        partition<T>(
-            array: Array<T>,
-            iterator: _.ListIterator<T, boolean>,
-            context?: any): T[][];
+         * Splits `collection` into two arrays: one whose elements all satisfy
+         * `iteratee` and one whose elements all do not satisfy `iteratee`.
+         * @param collection The collection to partition.
+         * @param iteratee The iteratee that defines the partitioning scheme
+         * for each element in `collection`.
+         * @param context `this` object in `iteratee`, optional.
+         * @returns An array composed of two elements, where the first element
+         * contains the elements in `collection` that satisfied the predicate
+         * and the second element contains the elements that did not.
+         **/
+        partition<V extends Collection<any>>(
+            list: V,
+            iteratee: Iteratee<V, boolean>,
```

Comment:
I considered making `iteratee` optional for `partition` to support the identity iteratee, but when I went to write tests for that they didn't seem super useful to me (basically sticking falsy values in one array and truthy values in another). Do folks feel that making this optional would be preferable, and if so would it be worth making a separate overload that attempts to put only potentially truthy types in the true tuple element and only potentially falsy types in the false tuple element?

Reply:
Making it optional is more truthful and not much work. I also wouldn't rule out partitioning by truthyness  as something that somebody might find useful, just like with filtering by truthyness.

As for narrowing the types of the tuple elements, that might actually unlock interesting new use cases since the separation of types would be very sharp. Up to you whether you find that worth the effort, though.

Reply:
I'm already planning on introducing a `NonFalsy` type for `compact`, might as well add that and a `NonTruthy` type as well.

Reply:
Actually I'll reverse that and make them `Truthy` and `Falsy` 😄

Reply:
Regrettably while I can filter out `0` and `''` specifically from `Truthy`, I don't think there's a way for me to filter out all numbers but those in `Falsy` unless I also exclude `string` and `number`, which I can't do, so they're unfortunately slightly lopsided when it comes to constant string and number values. I could just let all falsy constants end up in both if it would be preferable to keep these symmetrical.

Actually since I can't reference `bigint` I don't think I can do `Falsy` correctly, so I think it's probably best if I abandon this.

Reply:
Wait, how about something like this?

```ts
type AnyFalsy = void | null | false | 0 | -0 | NaN | '';

type Truthy<T> = Exclude<T, AnyFalsy>;
type Falsy<T> = Extract<T, AnyFalsy>;

// return type of _.partition with identity iteratee
[Truthy<TypeOfCollection<V>>[], Falsy<TypeOfCollection<V>>[]]
```

This omits `0n` and `document.all`, but I wouldn't feel guilty about either.

Reply:
The issue with what I believe to be a correct `Falsy` is the non-constant types like `number`, `string`, and `bigint`; if all I know is that I have a type that has constants that could be truthy or falsy, I would expect that type to show up in the results of both `Truthy` and `Falsy`.

So for this to be correct from my perspective, the following checks need to pass for all such types (using `bigint` as an example since it's the most problematic):

```TS
declare const truthyBigint: Truthy<bigint>; // $ExpectType bigint
declare const falsyBigint: Falsy<bigint>; // $ExpectType bigint
```

This presents two problems: 1) truthy constants like `1` and `'false'` are matched by non-constants like `number` and `string` and 2) `Falsy` needs to be able to extract `bigint` but I can't actually refer to `bigint` directly without supporting multiple versions of types.

Reply:
Don't you think this is much more useful and informative? And in fact, more correct?

```ts
declare const truthyBigint: Truthy<bigint>; // $ExpectType Exclude<bigint, 0n>
declare const falsyBigint: Falsy<bigint>; // $ExpectType 0n
```

Now, I didn't include `0n` in `AnyFalsy`, but otherwise I think the above would work.

Bigint is a problem if you don't want to upgrade to TS 3.2. In that case, I suggest ignoring it, because it  doesn't have nearly enough browser support to be usable in production yet. I don't see any problem with the other types that are involved.

Reply:
Regrettably with the types you outlined `Falsy<number>` results in `never` because `Extract<number, 0>` results in `never`. That does not provide an accurate falsy type because `number` includes `0`. I am not willing to commit or approve any solution in which `Falsy<number>` does not result in `number` (and the same goes for `Falsy<string>` due to `''`).

With regards to `bigint`, I agree that I also personally don't care about `bigint` (at this time at least), but even if the vast majority of developers also agree with that sentiment I worry that a solution that doesn't work with `bigint` also won't work well for any type introduced in the future that includes both falsy and truthy values and that may be more immediately popular, and I worry about adding a type like this that then potentially puts future developers between a rock (supporting multiple versions of types at once), a hard place (removing support for an established `Falsy` type), and...something else that goes with that analogy (pretending like the type doesn't exist and requiring folks to assert results until the minimum supported version creeps up to the version that introduced the type).

Reply:
Well, I guess to be 100% accurate I would be willing to go with a solution in which `Falsy<number>` resulted in some type like `!0` or `~0` or something similar, though there might not be an equivalent type that represents non-empty strings...

Reply:
I guess I was being too optimistic about TS using set-theoretic types. In Haskell, which TS often seems to take inspiration from, a type like `number` is just a (near-)infinite enum of all the possible values it can take. If TS behaved like that, `Extract<number, 0>` would just be `0`, which is exactly what I think it should be. Alas, with that type actually evaluating to `never`, that's of course not going to work.

I don't see the logic behind it evaluating to `never`, by the way.

Reply:
I found a way to do this that almost works correctly using a version of `Extract` that results in the second type instead of the first when the types match to get the wider type (e.g. `string`) when the narrower type (e.g. `0`) is specified. However, even though `object` by itself evaluates to `never` somehow specifically when it's included in a type union it ends up sticking around and I'm not sure why.

```TS
type AnyFalsy = undefined | null | false | 0 | '';

type BackwardsExtract<T, U> = T extends U ? U : never;

type Falsy<T> = BackwardsExtract<AnyFalsy, T>;

declare const falsyString: Falsy<string>; // results in string (when using Extract the result is '')
declare const falsyEmptyString: Falsy<''>; // results in ''
declare const truthy: Falsy<object>; // results in never
declare const allFalsies = Falsy<undefined | null | boolean | string | number | object | '' | 0 | false | true | 1 
| 'false'>; // results in undefined | null | boolean | string | number | object
```

It still doesn't deal with `bigint` though since `Falsy<bigint>` and `Falsy<0n`> still both result in `never`, which I'm not a huge fan of. Admittedly `compact` already uses some truthy logic that doesn't drop the `0n` constant from

Reply:
Using that `any` trick to spread out the type union `U` in advance seems to provide the desired behavior (and interestingly even somehow figures out when `boolean` is provided that it should only include `false` in its possible results, though looking at `Truthy` it seems to do the same thing).

```TS
type AnyFalsy = undefined | null | false | 0 | '';

type BackwardsExtract<T, U> = U extends any ? T extends U ? U : never : never;

type Falsy<T> = BackwardsExtract<AnyFalsy, T>;

declare const falsyString: Falsy<string>; // results in string (when using Extract the result is '')
declare const falsyConstant: Falsy<''>; // results in ''
declare const truthy: Falsy<object>; // results in never
declare const truthyConstant: Falsy<1>; // results in never
declare const allFalsies = Falsy<undefined | null | boolean | string | number | object | '' | 0 | 1 | 'false' | StringRecord>; // results in undefined | null | false | string | number
```

Reply:
Super extra silly me, here I was being all happy that I figured out the thing that would get me `number` when `number` was specified when you're right, it would be more accurate to get `0` rather than `number`...except then as I was writing that sentence I realized that it wasn't true because of `NaN` which regrettably doesn't have its own constant type like every other number (the TS compiler gets angry when `NaN` is used as a type and unlike every other constant number the inferred type of `const num = NaN;` is `number`).

Al that being said, if `NaN` didn't exist the below, which is almost exactly what you put together (the types provided to `Extract` are just reversed), would be even more correct than the use of `BackwardsExtract` and much simpler.

```TS
type AnyFalsy = undefined | null | false | 0 | '';

type Falsy<T> = Extract<AnyFalsy, T>;

declare const falsy: Falsy<undefined | null | boolean | string | number | object | '' | 0 | 1 | 'false' | StringRecord>;

declare const falsyString: Falsy<string>; // results in ''
declare const falsyConstant: Falsy<''>; // results in ''
declare const truthy: Falsy<object>; // results in never
declare const truthyConstant: Falsy<1>; // results in never
declare const allFalsies = Falsy<undefined | null | boolean | string | number | object | '' | 0 | 1 | 'false' | StringRecord>; // results in AnyFalsy
```

Reply:
<sigh> All of these solutions still completely exclude all `bigint` types, though, which still worries me more than `Truthy`'s behavior of just failing to filter out the `0n` constant specifically, so even though I'm glad we thought through all of this overall I'd still rather not do any of these.

Also sorry for somehow getting things completely backwards in my head at one point, re-reading what I was saying regarding `Falsy<number>` ideally resulting in something like `~0` and some type that is all strings but empty I'm not sure what I was thinking since that would be the exact opposite of correct 😞.

Reply:
I was considering creating a suggestion for `Truthy` and `Falsy` types in TS core type library files as a way to be able to deal with any new types like `bigint` with truthy and falsy constant values that may be introduced in the future since that seems like a useful way to deal with the `bigint` issue, particularly since TS already seems to have a general concept of what's truthy and what's falsy since type guards take away falsy types. However, I didn't end up doing so because when searching for existing issues I found a few that were of interest and even an open PR for adding such types that appears to have been backburnered in September 2018 but was possibly at least briefly noticed again and possibly brought back into active consideration in January. Links to those issues and the PR I mentioned are below.

microsoft/TypeScript#15135 - [Issue] NaN, Infinity and -Infinity not accepted in number literal types
microsoft/TypeScript#36964 - [Issue] NaN is not recognized as falsy
microsoft/TypeScript#27994 - [Issue] Design Meeting notes, 10/19/2018
microsoft/TypeScript#27050 - [Issue] Type parameter should not be assumed truthy
microsoft/TypeScript#22348 - [PR] Use lib conditional types for type facts if possible

Reply:
To wrap all of this up:

There are a few different solutions that could be adopted here that are somewhat but not completely correct, and I have more hope than I did when originally thinking through this that future cases similar to `bigint` may be dealt with by the introduction of `Truthy` and `Falsy` types in TS core type definitions. Since there isn't a home-run solution here at this time, I'd rather not introduce these types in this PR, but I'd be open to revisiting this in a separate future PR near the end of this project that applies these types to all the places where they might be useful and which will make a merge commit that can more easily be reverted wholesale if any major issues are reported or the lack of `bigint` support is more damaging than anticipated (assuming that the PR isn't created at a time where the minimum version is still below TS3.2).

Reply:
I was very confused when you suggested that `Falsy<number>` should be `number`. Glad that you agreed in the end that this was a very confusing position to take. 😅 

Very instructive that `Extract` apparently works in the opposite direction of what I thought. How about `Exclude` in `Truthy`, though?

I would be happy enough with a "best possible" approach that doesn't take `NaN` and `bigint` into account, but I'm also fine with postponing this to a later PR.

Reply:
The usage of `Exclude` seems to be correct, and having given some thought to all of this I do think the reversal between the two makes sense.

As a less technical explanation for why I think this all makes sense, I'd say that for `Truthy` we're trying to exclude falsy types from the set we're given (so the types we really want to end up with are all in `T` and that's what we exclude from) but for `Falsy` we're trying to arrive at as many of the falsy types as are included in what we're given (so the types we really want to end up with are all in `AnyFalsy` and so that's what we extract from).

As a more technical explanation for why I think this all makes sense, these types both use `extends` to figure out what should be included and excluded using polymorphism, and in general the way the relationships seem to work for `0` and `number` is that `0` is a more specific type of `number` similar to how `Car` is a more specific type of `Vehicle`. To take advantage of that, in the truthy case we want to only exclude `0` specifically without excluding all `number`s, so we want `0` to be on the base (right) side of the `extends` clause and for the left side to be the result if the check fails so we keep `number` around. On the other hand, in the `falsy` case we want to get `0` for any type that could include it, so we want `0` to be on the extending (left) side of the `extends` clause and to be the result if the check succeeds so `number` can be turned into `0`.

Reply:
I think I follow you and it seems to make sense. Thanks!